randovania/generator/hint_distributor.py

```diff
@@ -1,50 +1,87 @@
 from __future__ import annotations
 
-import copy
 import dataclasses
+import math
 from abc import ABC
-from collections.abc import Callable, Iterator
+from collections import defaultdict
+from collections.abc import Callable, Collection
 from random import Random
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Any, TypeVar
 
-from randovania.game_description import node_search
 from randovania.game_description.db.hint_node import HintNode, HintNodeKind
 from randovania.game_description.db.pickup_node import PickupNode
 from randovania.game_description.game_patches import GamePatches
 from randovania.game_description.hint import (
+    PRECISION_PAIR_UNASSIGNED,
     HintItemPrecision,
     HintLocationPrecision,
-    HintRelativeAreaName,
     JokeHint,
     LocationHint,
     PrecisionPair,
-    RelativeData,
-    RelativeDataArea,
-    RelativeDataItem,
     is_unassigned_location,
 )
 from randovania.game_description.resources.pickup_index import PickupIndex
 from randovania.generator.filler.filler_library import UnableToGenerate
-from randovania.generator.filler.player_state import PlayerState
-from randovania.lib import random_lib
+from randovania.generator.filler.player_state import HintState, PlayerState
 from randovania.resolver import debug
 
 if TYPE_CHECKING:
-    from randovania.game_description.db.area import Area
     from randovania.game_description.db.node_identifier import NodeIdentifier
     from randovania.game_description.db.region_list import RegionList
+    from randovania.game_description.hint_features import HintFeature, PickupHintFeature
     from randovania.game_description.pickup.pickup_entry import PickupEntry
     from randovania.generator.filler.filler_configuration import PlayerPool
     from randovania.generator.pre_fill_params import PreFillParams
 
 HintProvider = Callable[[PlayerState, GamePatches, Random, PickupIndex], LocationHint | None]
 
-
-def _not_empty(it: Iterator) -> bool:
-    return sum(1 for _ in it) > 0
+HintTargetPrecision = tuple[PickupIndex, PrecisionPair]
 
 
-HintTargetPrecision = tuple[PickupIndex, PrecisionPair]
+FeatureT = TypeVar("FeatureT")
+PrecisionT = TypeVar("PrecisionT")
+
+
+def choose_feature(
+    elements_with_feature: dict[FeatureT | PrecisionT, Collection[Any]],
+    total_elements: int,
+    detailed_precision: PrecisionT,
+    element_features: Collection[FeatureT],
+    rng: Random,
+    mean: float,
+    std_dev: float,
+) -> FeatureT | PrecisionT:
+    # arbitrarily increased until it felt good
+    DEGREE = 30
+    feature_precisions = {
+        feature: math.pow((total_elements - len(elements_with_feature[feature])) / (total_elements - 1), DEGREE)
+        for feature in elements_with_feature
+    }
+    feature_precisions = {
```

Comment:
I wonder if this could be combined with the above?

Reply:
I think it's a bit clearer what it's doing to do it in two steps